File: _CodeTopics/LeetCode/201-400/000300/000300_algo2_impl_newlib.py
import bisect
class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        # 同样是官方答案贪心 + 二分查找方法的思路，只不过这次用库实现二分查找部分。

        length = len(nums)
        d = [nums[0]]

        for i in range(1, length):
            if nums[i] > d[-1]:
                d.append(nums[i])
            else:
                loc = bisect.bisect_left(d, nums[i])
                d[loc] = nums[i]
        return len(d)
        
"""
https://leetcode-cn.com/submissions/detail/151173818/

54 / 54 个通过测试用例
状态：通过
执行用时: 36 ms
内存消耗: 13.3 MB

执行用时：36 ms, 在所有 Python 提交中击败了93.89%的用户
内存消耗：13.3 MB, 在所有 Python 提交中击败了55.10%的用户
"""

File: _CodeTopics/LeetCode/201-400/000300/tran/000300_algo2_impl_newlib.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {

        int length = nums.size();
        vector<int> d = {nums[0]};

        for (int i = 1; i < length; ++i) {
            if (nums[i] > d.back()) {
                d.push_back(nums[i]);
            } else {
                std::vector<int>::iterator loc = lower_bound(d.begin(), d.end(), nums[i]);
                // 也可以用 *loc = nums[i];
                d[loc - d.begin()] = nums[i];
            }
        }
        return d.size();
    }
};

/*
https://leetcode-cn.com/submissions/detail/257496079/

执行用时：8 ms, 在所有 C++ 提交中击败了90.55%的用户
内存消耗：10.2 MB, 在所有 C++ 提交中击败了66.86%的用户
通过测试用例：
54 / 54
*/
